docs/design/prerequisites-and-targets.md

```diff
@@ -0,0 +1,117 @@
+# Prerequisites and targets
+
+## Summary
+
+This presents a design for two mechanisms that can be combined to solve the "stale provider state"
+problem described in [#299](https://github.com/pulumi/pulumi-kubernetes-operator/issues/299).
+
+## Motivation
+
+Some Pulumi programs are structured in phases, such that inputs obtained in phase `N` are used to
+construct a _provider_ for creating resources in phase `N+1`. This presents a problem for the Pulumi
+machinery, because inputs to the provider must be kept in the stack state, and will become stale if
+the resource in question changes; e.g., if they are credentials and are rotated. Issuing a `pulumi
+refresh` (or `pulumi up -r`, or Stack object with `.spec.refresh` set) will balk in that situation,
+because refreshing consults the state rather than running the program, so it will proceed with stale
+credentials and likely fail authentication.
+
+Usually it is possible to work around this by using `pulumi up -t <provider URN>` to rerun the
+program and update the inputs to the provider, before running the refresh operation. However, this
+workaround isn't available when using the operator, because there is no representation of "run this
+before" or targeted updates in its API.
+
+## Design
+
+This design involves two additions to the API:
+
+ - "prerequisites", a list of other stack objects that must be run before the stack under
+   consideration is run; and,
+ - "targets", a list of URNs to be given to the Pulumi engine when running the stack.
+
+### Prerequisites
+
+The field `.spec.prerequisites` gives a list of references to stacks. The controller makes sure each
+has been run recently before running the stack under consideration. This means, for instance, you
+can update credentials in the state before re-running a program that uses those credentials.
+
+Usually it will be important that a prerequisite has run recently. The field
+`.spec.prerequisite[].updatedWithinDuration` puts a limit on how long ago it was last run:
+
+```
+spec:
+  prerequisites:
+  - stackRef:
+      name: update-credentials
+    unless:
+      updatedWithinDuration: 1m
+```
+
+If the referenced stack has not been run in the last `updatedWithinDuration` duration, it will be
+queued, and the current stack requeued.
+
+**Q. How does the controller queue the prerequisites?**
+
+One mechanism is to simply add an annotation to the object, which will trigger an update event and
```

Comment:
If we are using the drift-detection mode (`ContinueResyncOnCommitMatch`) we wouldn't need to tickle the stack right? I imagine we could align this with `updatedWithinDuration`.

Reply:
Right, you could just let the prerequisite get run in the normal course of things, and come back later. There's a chance the schedules won't match well, and you get syncopated updates -- that is, you sometimes or always miss your `updatedWithinDuration` window because the resync period is too long. In that case, explicitly queueing prerequisites may be necessary for correctness.